sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Triggers/EventHubAsyncCollector.cs

```diff
@@ -38,19 +39,40 @@ public EventHubAsyncCollector(IEventHubProducerClient client)
         }
 
         /// <summary>
-        /// Add an event.
+        /// Initializes a new instance of the <see cref="EventHubAsyncCollector" /> class.
+        /// </summary>
+        /// <remarks>This constructor is intended to be used for mocking scenarios only.</remarks>
+        protected EventHubAsyncCollector()
+        {
+        }
+
+        /// <summary>
+        /// Add an event to be published with round-robin partition assignment.
         /// </summary>
         /// <param name="item">The event to add</param>
-        /// <param name="cancellationToken">a cancellation token. </param>
-        /// <returns></returns>
-        public async Task AddAsync(EventData item, CancellationToken cancellationToken = default(CancellationToken))
+        /// <param name="cancellationToken">A token that can be used to cancel the operation.</param>
+        public virtual Task AddAsync(EventData item, CancellationToken cancellationToken = default(CancellationToken)) => AddAsync(item, null, cancellationToken);
+
+        /// <summary>
+        /// Add an event to be published using the provided <paramref name="partitionKey"/> for partition assignment.
+        /// </summary>
+        /// <param name="item">The event to add</param>
+        /// <param name="partitionKey">The partition key to use for partition assignment.  If <c>null</c>, round-robin partition assignment will be used.</param>
+        /// <param name="cancellationToken">A token that can be used to cancel the operation.</param>
+        public virtual async Task AddAsync(EventData item, string partitionKey, CancellationToken cancellationToken = default(CancellationToken))
         {
             if (item == null)
             {
                 throw new ArgumentNullException(nameof(item));
             }
 
-            string key = item.PartitionKey ?? string.Empty;
+            // If the partition key has no value, set it to null.
+            if (partitionKey is { Length: 0 })
```

Comment:
Interesting.. why do we need to set to null if we are just resetting to empty on L75?

Reply:
That is an excellent question to which I have no answer.   I assume that I was doing something with the null at one point - but it sure doesn't add any value now.

Reply:
Spoke too soon.   We need to have `partitionKey` null (and not empty) for L86 when we create the batch, as an empty string is a valid partition key where `null` signals that no key is present.   The `key` is used as a dictionary key, which must be non-null.